File: frontend/utils/productLocalization.ts
import type { Product } from '../services/api/productService';

export type ProductTextLocale = 'de' | 'en' | 'tr';

export type LocalizedProductFields = {
  name?: string;
  nameDe?: string | null;
  nameEn?: string | null;
  nameTr?: string | null;
  description?: string | null;
  descriptionDe?: string | null;
  descriptionEn?: string | null;
  descriptionTr?: string | null;
};

export function normalizeProductTextLocale(language: string | undefined): ProductTextLocale {
  const base = (language ?? 'de').split('-')[0].toLowerCase();
  if (base === 'en' || base === 'tr') return base;
  return 'de';
}

function firstNonEmpty(...values: Array<string | null | undefined>): string {
  for (const value of values) {
    if (value != null && String(value).trim() !== '') {
      return String(value).trim();
    }
  }
  return '';
}

function firstNonEmptyOptional(...values: Array<string | null | undefined>): string | undefined {
  const hit = firstNonEmpty(...values);
  return hit || undefined;
}

export function resolveProductDisplayName(
  product: LocalizedProductFields,
  locale?: string,
): string {
  const lang = normalizeProductTextLocale(locale);
  switch (lang) {
    case 'en':
      return firstNonEmpty(product.nameEn, product.nameDe, product.name);
    case 'tr':
      return firstNonEmpty(product.nameTr, product.nameDe, product.name);
    default:
      return firstNonEmpty(product.nameDe, product.name);
  }
}

export function resolveProductDisplayDescription(
  product: LocalizedProductFields,
  locale?: string,
): string | undefined {
  const lang = normalizeProductTextLocale(locale);
  switch (lang) {
    case 'en':
      return firstNonEmptyOptional(product.descriptionEn, product.descriptionDe, product.description);
    case 'tr':
      return firstNonEmptyOptional(product.descriptionTr, product.descriptionDe, product.description);
    default:
      return firstNonEmptyOptional(product.descriptionDe, product.description);
  }
}

/** View-model with locale-resolved name/description for POS lists. */
export function withLocalizedProductDisplay<T extends LocalizedProductFields>(
  product: T,
  locale?: string,
): T & { displayName: string; displayDescription?: string } {
  return {
    ...product,
    displayName: resolveProductDisplayName(product, locale),
    displayDescription: resolveProductDisplayDescription(product, locale),
  };
}

export function productMatchesSearchQuery(product: Product, query: string, locale?: string): boolean {
  const q = query.trim().toLowerCase();
  if (!q) return true;
  const names = [
    product.name,
    product.nameDe,
    product.nameEn,
    product.nameTr,
    resolveProductDisplayName(product, locale),
  ];
  return names.some((n) => n?.toLowerCase().includes(q));
}

File: backend/Services/ProductLocalization.cs
using KasseAPI_Final.Models;

namespace KasseAPI_Final.Services;

/// <summary>Resolves localized product display strings for POS/catalog (de, en, tr).</summary>
public static class ProductLocalization
{
    public static void SyncCanonicalFields(Product product)
    {
        product.Name = ResolveName(product, "de");
        product.Description = ResolveDescription(product, "de");
    }

    public static string ResolveName(Product product, string? locale)
    {
        return locale?.Split('-')[0].ToLowerInvariant() switch
        {
            "en" => FirstNonEmpty(product.NameEn, product.NameDe, product.Name),
            "tr" => FirstNonEmpty(product.NameTr, product.NameDe, product.Name),
            _ => FirstNonEmpty(product.NameDe, product.Name),
        };
    }

    public static string? ResolveDescription(Product product, string? locale)
    {
        return locale?.Split('-')[0].ToLowerInvariant() switch
        {
            "en" => FirstNonEmptyOptional(product.DescriptionEn, product.DescriptionDe, product.Description),
            "tr" => FirstNonEmptyOptional(product.DescriptionTr, product.DescriptionDe, product.Description),
            _ => FirstNonEmptyOptional(product.DescriptionDe, product.Description),
        };
    }

    public static void ApplyLocalizedNames(
        Product product,
        string? nameDe,
        string? nameEn,
        string? nameTr,
        string? descriptionDe = null,
        string? descriptionEn = null,
        string? descriptionTr = null)
    {
        if (!string.IsNullOrWhiteSpace(nameDe))
            product.NameDe = nameDe.Trim();
        if (!string.IsNullOrWhiteSpace(nameEn))
            product.NameEn = nameEn.Trim();
        if (!string.IsNullOrWhiteSpace(nameTr))
            product.NameTr = nameTr.Trim();
        if (descriptionDe != null)
            product.DescriptionDe = string.IsNullOrWhiteSpace(descriptionDe) ? null : descriptionDe.Trim();
        if (descriptionEn != null)
            product.DescriptionEn = string.IsNullOrWhiteSpace(descriptionEn) ? null : descriptionEn.Trim();
        if (descriptionTr != null)
            product.DescriptionTr = string.IsNullOrWhiteSpace(descriptionTr) ? null : descriptionTr.Trim();

        SyncCanonicalFields(product);
    }

    private static string FirstNonEmpty(params string?[] values)
    {
        foreach (var value in values)
        {
            if (!string.IsNullOrWhiteSpace(value))
                return value.Trim();
        }

        return string.Empty;
    }

    private static string? FirstNonEmptyOptional(params string?[] values)
    {
        foreach (var value in values)
        {
            if (!string.IsNullOrWhiteSpace(value))
                return value.Trim();
        }

        return null;
    }
}
